### PR TITLE
Create incremental snapshot from full rollback hash

### PR DESCRIPTION
**Pull request by kriscieslak**

### CODE REVIEW

File: modules/core/src/main/scala/org/tessellation/Main.scala [hunk 1]
```diff
@@ -126,25 +127,45 @@ object Main
...
-                  )
-                snapshotTraverse.loadChain().flatMap {
-                  case (snapshotInfo, snapshot) =>
-                    storages.globalSnapshot.prepend(snapshot, snapshotInfo) >>
-                      services.collateral
-                        .hasCollateral(sdk.nodeId)
-                        .flatMap(OwnCollateralNotSatisfied.raiseError[IO, Unit].unlessA) >>
-                      services.consensus.manager.startFacilitatingAfter(snapshot.ordinal, snapshot, snapshotInfo)
-                }
+            SnapshotLocalFileSystemStorage.make[IO, GlobalSnapshot](cfg.snapshot.snapshotPath).flatMap {
```
Comment (marcinwadon) [inline modules/core/src/main/scala/org/tessellation/Main.scala:130]: IMO if it's so large, it shouldn't be a part of the Main file, but a Program

File: modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/storage/SnapshotStorage.scala [hunk 2]
```diff
@@ -124,14 +124,22 @@ object SnapshotStorage {
         hashCache(hash).set(snapshot.some) >>
           ordinalCache(snapshot.ordinal).set(hash.some) >>
           snapshotLocalFileSystemStorage.write(snapshot).handleErrorWith { e =>
-            logger.error(e)(s"Failed writing snapshot to disk! hash=$hash ordinal=${snapshot.ordinal}") >>
-              notPersistedCache.update(current => current + snapshot.ordinal)
+            snapshotExists(snapshot).ifM(
+              logger.info(s"Snapshot is already saved on disk. hash=$hash ordinal=${snapshot.ordinal}"),
+              logger.error(e)(s"Failed writing snapshot to disk! hash=$hash ordinal=${snapshot.ordinal}") >>
+                notPersistedCache.update(current => current + snapshot.ordinal)
+            )
```
Comment (marcinwadon) [inline modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/storage/SnapshotStorage.scala:131]: It checks whether snapshot exists under the hash, but it should exist under the ordinal as well, so either update `snapshotExists` to check both hash and ordinal, or update `notPersistedCache` as it was before.

**Comment by marcinwadon** [comment]
also, label is missing

File: modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/storage/SnapshotStorage.scala (hunk 2 above)
Comment (marcinwadon) [reply modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/storage/SnapshotStorage.scala:131]: bump @kriscieslak

**Review by marcinwadon — CHANGES_REQUESTED** [review]
LGTM, but please answer to the comment re: persisted cache update

File: modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/storage/SnapshotStorage.scala (hunk 2 above)
Comment (kriscieslak) [reply modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/storage/SnapshotStorage.scala:131]: I have improved snapshotExists method to check by hash and ordinal name.

**Review by marcinwadon — APPROVED** [review]

File: modules/core/src/main/scala/org/tessellation/modules/Programs.scala [hunk 3]
```diff
@@ -1,33 +1,48 @@
...
+import org.tessellation.kryo
 import org.tessellation.sdk.domain.cluster.programs.{Joining, PeerDiscovery}
 import org.tessellation.sdk.infrastructure.snapshot.programs.Download
 import org.tessellation.sdk.modules.SdkPrograms
+import org.tessellation.security.SecurityProvider
 
 object Programs {
 
-  def make[F[_]: Async](
+  def make[F[_]: Async: kryo.KryoSerializer: SecurityProvider](
```
Comment (marcinwadon) [inline modules/core/src/main/scala/org/tessellation/modules/Programs.scala:18]: import

**Review by marcinwadon — DISMISSED** [review]

**Review by marcinwadon — APPROVED** [review]

**Review by marcinwadon — APPROVED** [review]

**Pull request merged**